### PR TITLE
x86_64-linux: use shared libraries instead of static libraries

### LINKED ISSUE #54: Add a Nix overlay that provides llc and wasm-ld executables
For compiling to WebAssembly, we need the `llc` (from LLVM14 toolchain) and `wasm-ld` executables.

A small overlay (called `wasm-toolchain-overlay`?) that exposes those 2 executables should be added to the `flake.nix`.

Additionally, we can add it [here](https://github.com/luc-tielen/eclair-lang/blob/main/flake.nix#L66) so the tools are automatically available during development.

Finally, we can also expose it to other flakes just like we do for the [main eclair overlay](https://github.com/luc-tielen/eclair-lang/blob/main/flake.nix#L74). This would probably be a simple `inherit wasm-toolchain-overlay;`?

### PR DESCRIPTION
**Pull request by smunix**
fix #54 , #14 
Also use `nixfmt` to format `flake.nix` file

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -5,99 +5,129 @@
...
-              shs.overlay."${system}"
-              llvm-cg.overlay."${system}"
-              overlay
+                    diagnose =
+                      hf.callCabal2nixWithOptions "diagnose" (inputs.diagnose)
+                      "-fmegaparsec-compat" { };
+
+                    eclair-lang = with hf;
+                      with final.haskell.lib;
+                      addExtraLibraries (disableLibraryProfiling
```
Comment (luc-tielen) [inline flake.nix:74]: Can we turn this into a helper function or something?

File: flake.nix [hunk 2]
```diff
@@ -5,99 +5,129 @@
...
-                        });
-                    };
-                };
-            in { inherit haskellPackages; };
-          pkgs = import np {
-            inherit system config;
-            overlays = [
-              ds.overlay
-              shs.overlay."${system}"
-              llvm-cg.overlay."${system}"
```
Comment (luc-tielen) [inline flake.nix:69]: Too bad we couldn't fix this in the llvm-codegen overlay. :/

**Review by luc-tielen — CHANGES_REQUESTED** [review]
Overall looks good. But left some questions.

Also a tip for next time: run a formatter in a separate commit, so rest is easier to track.

File: flake.nix [hunk 3]
```diff
@@ -5,99 +5,129 @@
...
-            let
-              haskellPackages =
-                final.haskell.packages."ghc${ghcVersion}".override {
-                  overrides = hf: hp:
-                    with final.haskell.lib; rec {
-                      inherit (shs.packages."${system}") souffle-haskell;
-                      inherit (llvm-cg.packages."${system}") llvm-codegen;
+    in with np.lib;
+    with fu.lib;
+    eachSystem [ "x86_64-linux" ] (system:
```
Comment (luc-tielen) [inline flake.nix:27]: I think we still need to add x86_64-darwin to actually make it run on that arch.

File: flake.nix [hunk 4]
```diff
@@ -5,99 +5,129 @@
     # np.url = "github:nixos/nixpkgs?ref=haskell-updates";
     fu.url = "github:numtide/flake-utils?ref=master";
     ds.url = "github:numtide/devshell?ref=master";
+    nf.url = "github:numtide/nix-filter?ref=master";
     hls.url = "github:haskell/haskell-language-server?ref=master";
     shs.url =
       "github:luc-tielen/souffle-haskell?rev=c46d0677e4bc830df89ec1de2396c562eb9d86d3";
-    llvm-cg.url =
-      "github:luc-tielen/llvm-codegen?rev=83306c9da879b80be674ef7cf910aaf158160b69";
+    llvm-cg.url = "github:luc-tielen/llvm-codegen";
```
[the remark this answers is not in the document]
Comment (smunix) [reply flake.nix:12]: Correct, it's still pinned. `nix flake lock --update-input llvm-cg` will update it accordingly, if needed.

You can also use the "?rev=commits" to point to a specific one of you need to.

File: flake.nix [hunk 5]
```diff
@@ -5,99 +5,129 @@
...
+            haskellPackages =
+              final.haskell.packages."ghc${ghcVersion}".override {
+                overrides = hf: hp:
+                  with final.haskell.lib; rec {
+                    inherit (shs.packages."${system}") souffle-haskell;
 
-                      algebraic-graphs = with hf;
-                        dontCheck
-                        (callCabal2nix "algebraic-graphs" (inputs.alga) { });
+                    llvm-codegen = addExtraLibraries (enableSharedExecutables
```
[the remark this answers is not in the document]
Comment (smunix) [reply flake.nix:47]: I have a better helper function, mostly for composition, which I can push here if you'd like

File: flake.nix (hunk 3 above)
Comment (smunix) [reply flake.nix:27]: Yes, that's the next step indeed. And that's part of the reason why I'm removing the llvm-cg overlay. @placidex has successfully tested it in a separate branch

**Review by smunix — COMMENTED** [review]
@luc-tielen , this is good for review/merge, I don't intend on making further changes. I created a utility function `mkCabal` in `github:smunix/nix-utils` to simplify cabal package creation.

File: flake.nix [hunk 6]
```diff
@@ -5,99 +5,129 @@
...
     fu.url = "github:numtide/flake-utils?ref=master";
     ds.url = "github:numtide/devshell?ref=master";
+    nf.url = "github:numtide/nix-filter?ref=master";
     hls.url = "github:haskell/haskell-language-server?ref=master";
     shs.url =
       "github:luc-tielen/souffle-haskell?rev=c46d0677e4bc830df89ec1de2396c562eb9d86d3";
-    llvm-cg.url =
-      "github:luc-tielen/llvm-codegen?rev=83306c9da879b80be674ef7cf910aaf158160b69";
+    llvm-cg.url = "github:luc-tielen/llvm-codegen";
+    llvm-cg.flake = false;
```
[the remark this answers is not in the document]
Comment (luc-tielen) [reply flake.nix:13]: Ok.

File: flake.nix (hunk 3 above)
Comment (luc-tielen) [reply flake.nix:28]: Ok, didn't realize that!

**Review by luc-tielen — APPROVED** [review]

File: flake.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (luc-tielen) [reply flake.nix]: Yeah I looked at it, let's use it directly from your nix-utils overlay. Thanks!

**Comment by luc-tielen** [comment]
Tried building it locally before merging, it's broken?

```
error: flake 'git+file:///tmp/eclair-lang' does not provide attribute 'devShells.x86_64-linux.devShell.x86_64-linux', 'packages.x86_64-linux.devShell.x86_64-linux', 'legacyPackages.x86_64-linux.devShell.x86_64-linux', 'devShell.x86_64-linux' or 'defaultPackage.x86_64-linux'
```

Both `nix build` and `nix develop` don't work because of this.

File: flake.nix [hunk 7]
```diff
@@ -5,99 +5,131 @@
...
             # Next line always sets DATALOG_DIR so souffle can find the datalog files in interpreted mode.
-            env = [{ name = "DATALOG_DIR"; value = "cbits/"; }];
+            env = [{
+              name = "DATALOG_DIR";
+              value = "cbits/";
+            }];
           };
-        });
+        };
+      in { inherit overlays packages devShells; });
```
Comment (luc-tielen) [inline flake.nix:134]: @smunix I think you made a mistake here? Because some keys have a ".default" key in between..

**Review by luc-tielen — CHANGES_REQUESTED** [review]
Latest refactor broke the setup

File: flake.nix (hunk 7 above)
Comment (smunix) [reply flake.nix:134]: No mistake, you need to update your local nix version. See the changes I made earlier in the GitHub action `build.yaml` file?
The current stake version of Nix is `2.11.0`, yours is certainly way behind. Can you check?

**Comment by smunix** [comment]
> Latest refactor broke the setup

No, is not broken. Update your local nix (see the CI change here with `build.yaml`)

**Review by luc-tielen — APPROVED** [review]

**Pull request merged**